zsjed: init at 0.3

fogti opened the pull request:
<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!
List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

###### Motivation for this change

Packaging my semi-stenography tool (hiding file snippets at the end of other files).

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [X] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux) (enabled by default)
- Built on platform(s)
   - [X] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [X] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [X] Ensured that relevant documentation is up to date
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

SuperSandro2000 commented on maintainers/maintainer-list.nix:
```diff
@@ -10686,6 +10686,16 @@
     github = "pulsation";
     githubId = 1838397;
   };
+  zseri = {
```
Please put this into a separate commit with the message ``maintainers: add zseri``.

SuperSandro2000 commented on pkgs/tools/misc/zsjed/default.nix:
```diff
@@ -0,0 +1,62 @@
+  '';
+
+  meta = with lib; {
+    # https://github.com/zserik/portage-zscheile/blob/ea553888fff7f9728d60082e224fd808f2a8924d/media-metad/zsjed/zsjed-0.3.ebuild
```
```suggestion
```

SuperSandro2000 commented on pkgs/tools/misc/zsjed/default.nix:
```diff
@@ -0,0 +1,62 @@
+  meta = with lib; {
+    # https://github.com/zserik/portage-zscheile/blob/ea553888fff7f9728d60082e224fd808f2a8924d/media-metad/zsjed/zsjed-0.3.ebuild
+    description = "Zscheile data split semi-steganography tool";
+    # https://github.com/zserik/portage-zscheile/blob/ea553888fff7f9728d60082e224fd808f2a8924d/media-metad/zsjed/metadata.xml
```
```suggestion
```

SuperSandro2000 commented on pkgs/tools/misc/zsjed/default.nix:
```diff
@@ -0,0 +1,62 @@
+      It splits a file and pastes these parts
+      at the end of many other files, recoverable.
+    '';
+    # https://github.com/zserik/portage-zscheile/blob/ea553888fff7f9728d60082e224fd808f2a8924d/eclass/zs-minimal.eclass
```
```suggestion
```

SuperSandro2000 commented on pkgs/tools/misc/zsjed/default.nix:
```diff
@@ -0,0 +1,62 @@
+    install -t $out/bin zsjed-collect zsjed-clear zsjed-spread zsjed-mkls
+  '';
+
+  postFixup = let
+    runtimePath = lib.makeBinPath buildInputs;
+  in
+  ''
+    sed -i "2 i export PATH=${runtimePath}:\$PATH" $out/bin/zsjed-mkls
+  '';
```
```suggestion
  postFixup = ''
    sed -i "2 i export PATH=${lib.makeBinPath buildInputs}:\$PATH" $out/bin/zsjed-mkls
  '';
```

SuperSandro2000 commented on pkgs/tools/misc/zsjed/default.nix:
```diff
@@ -0,0 +1,62 @@
+
+  # the Makefile does not work for this, as it assumes the wrong prefix
+  installPhase = ''
+    mkdir -p $out/bin
+    install -t $out/bin zsjed-collect zsjed-clear zsjed-spread zsjed-mkls
```
```suggestion
    install -tD $out/bin zsjed-collect zsjed-clear zsjed-spread zsjed-mkls
```

zserik replied:
This doesn't work:
```
installing
install: failed to access 'D': No such file or directory
```

SuperSandro2000 replied:
I think the D needs to go before the t.

SuperSandro2000 commented on pkgs/tools/misc/zsjed/default.nix:
```diff
@@ -0,0 +1,62 @@
+    gawk
+  ];
+
+  # can't check because all tools return retcode 1 by default
```
We could work around this. Do you have tests?

zserik replied:
I found a workaround. Is it ok that the check has some output, or should I forward everything to `/dev/null`?

SuperSandro2000 commented on pkgs/tools/misc/zsjed/default.nix:
```diff
@@ -0,0 +1,62 @@
+    url = "https://ytrizja.de/distfiles/zsjed-${version}.tar.gz";
+    sha256 = "1rcrvrqjsv80ldz7cscqbrf98jb0yd2in97irrzlpask699z5k29";
+  };
+
+  nativeBuildInputs = [
+    coreutils # mkdir, install
+    gnused
+  ];
```
```suggestion
```
Should be included in stdenv.

zserik replied:
Is `diffutils` included in stdenv?

zserik replied:
I assume for now that is it not in stdenv.

SuperSandro2000 commented on pkgs/tools/misc/zsjed/default.nix:
```diff
@@ -0,0 +1,62 @@
+    gnused
+  ];
+
+  # runtime deps of zsjed-mkls
```
```suggestion
```

fogti committed:
maintainers: add zseri

fogti committed:
zsjed: init at 0.3

metadata taken from https://github.com/zserik/portage-zscheile/tree/ea553888fff7f9728d60082e224fd808f2a8924d/media-metad/zsjed

zserik commented on pkgs/tools/misc/zsjed/default.nix:
```diff
@@ -0,0 +1,68 @@
+  '';
+
+  postFixup = ''
+    sed -i "2 i export PATH=${lib.makeBinPath buildInputs}:\$PATH" $out/bin/zsjed-mkls
```
This is suboptimal, but I'm unsure how to use the `substitute*` fn's to put the appropriate paths in there...

SuperSandro2000 replied:
I think you might want to use makeWrapper here. You can take a look at somafm-cli for a simple example.

SuperSandro2000 commented on pkgs/tools/misc/zsjed/default.nix:
```diff
@@ -0,0 +1,68 @@
+
+  # doesn't test zsjed-mkls
+  checkPhase = ''
+    mkdir tmp
+    cd tmp
```
```suggestion
    tmp=$(mktemp -d)
    cd $tmp
```

SuperSandro2000 commented on pkgs/tools/misc/zsjed/default.nix:
```diff
@@ -0,0 +1,68 @@
+    ../zsjed-collect j l
+    cmp i j
+    cd ..
+    rm -rf tmp
```
```suggestion
```

zserik replied:
Huh, does Nix automatically remove any temp folder created by `mktemp -d`?

SuperSandro2000 replied:
nix-collect-garbage deletes it.

zserik commented on pkgs/tools/misc/zsjed/default.nix:
```diff
@@ -0,0 +1,68 @@
+    makeWrapper
+  ];
+
+  buildInputs = [
```
That `buildInputs` is handled equivalent to `depsHostTarget` is outright confusing... (see also: https://nixos.wiki/wiki/Cross_Compiling)

SuperSandro2000 replied:
Rule of thumb: everything linked against goes to buildInputs and everything executed or what should be in PATH goes to nativeBuildInputs. About all other cross compile issues I would just wait until someone discovers problems while building.

SuperSandro2000 commented on pkgs/tools/misc/zsjed/default.nix:
```diff
@@ -0,0 +1,68 @@
+    makeWrapper
+  ];
+
+  buildInputs = [
+    coreutils # sort
+    findutils
+    gawk
+  ];
+
```
```suggestion
```

zserik replied:
I guess, it's in stdenv, right?

zserik replied:
Is there a list of packages which are in stdenv?

SuperSandro2000 commented on pkgs/tools/misc/zsjed/default.nix:
```diff
@@ -0,0 +1,68 @@
+  # can't reuse the Makefile 'install' target, as it assumes the wrong prefix
+  installPhase = ''
+    install -D -t $out/bin zsjed-collect zsjed-clear zsjed-spread zsjed-mkls
+    wrapProgram $out/bin/zsjed-mkls --prefix PATH ":" "${lib.makeBinPath buildInputs}"
```
```suggestion
    wrapProgram $out/bin/zsjed-mkls --prefix PATH ":" "${lib.makeBinPath [ coreutils findutils gawk]}"
```

zserik commented:
Could I do anything to speed this process up?
(I think `nix.useSandbox` is enabled by default in NixOS, so that is already done; I'm unsure what "relevant documentation" in this case might be)

SuperSandro2000 commented:
> I think `nix.useSandbox` is enabled by default in NixOS, so that is already done

yes it is.

> I'm unsure what "relevant documentation" in this case might be

Not relevant for simple packages.

SuperSandro2000 commented:
This is a semi-automatic executed nixpkgs-review with [nixpkgs-review-checks extension](https://github.com/SuperSandro2000/nixpkgs-review-checks). It is checked by a human on a best effort basis and does not build all packages (e.g. lumo, tensorflow or pytorch).
If you have any questions or problems please reach out to SuperSandro2000 on IRC.

Result of `nixpkgs-review pr 112959` run on x86_64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>

  <summary>1 package failed to build and already failed to build on hydra master:</summary>
  <ul>
<li>zsjed: <a href="https://termbin.com/jy2o">plain log</a>  |  <a href="https://htmlpreview.github.io/?https://gist.githubusercontent.com/SuperSandroBot/68eca43f66cc1270fc6006b7dbe7bd00/raw/index.html">pretty log </a></li>
  </ul>
</details>

The following issues got detected with the above build packages.
Please fix at least the ones listed with your changed packages:

<details>
zsjed:
<details>


`checkPhase` should probably contain `runHook preCheck` and `runHook postCheck`.

Near pkgs/tools/misc/zsjed/default.nix:29:3:

```
   |
29 |   checkPhase = ''
   |   ^
```

See: https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/missing-phase-hooks.md
`installPhase` should probably contain `runHook preInstall` and `runHook postInstall`.

Near pkgs/tools/misc/zsjed/default.nix:46:3:

```
   |
46 |   installPhase = ''
   |   ^
```

See: https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/missing-phase-hooks.md
</details>
</details>

fogti committed:
zsjed: patch Makefile

fogti committed:
zsjed: fix invalid $out interpolation

zserik commented:
I'm probably going to port zsjed in a few weeks, and this program isn't critical, so I'm skipping this for now.
